core/src/main/java/io/questdb/std/ConcurrentAssociativeCache.java

```diff
@@ -0,0 +1,194 @@
+ * <p>
+ * The cache uses striped locking, i.e. each row is synchronized with its own lock.
+ */
+public class ConcurrentAssociativeCache<V> implements AssociativeCache<V> {
+    private static final int MIN_BLOCKS = 1;
+    private static final int MIN_ROWS = 1;
+    private final int blocks;
+    private final LongGauge cachedGauge;
+    private final Counter hitCounter;
+    // Each row has its own array of keys.
+    // Arrays are also used for locking.
+    private final ObjList<String[]> keys;
+    private final Counter missCounter;
+    private final int rowMask;
+    private final int rows;
+    // Each row has its own array of values.
+    private final ObjList<V[]> values;
+
+    public ConcurrentAssociativeCache(int blocks, int rows) {
+        this(blocks, rows, NullLongGauge.INSTANCE, NullCounter.INSTANCE, NullCounter.INSTANCE);
+    }
+
+    @SuppressWarnings("unchecked")
+    public ConcurrentAssociativeCache(int blocks, int rows, LongGauge cachedGauge, Counter hitCounter, Counter missCounter) {
+        this.blocks = Math.max(MIN_BLOCKS, Numbers.ceilPow2(blocks));
+        this.rows = Math.max(MIN_ROWS, Numbers.ceilPow2(rows));
+
+        int capacity = this.rows * this.blocks;
+        if (capacity < 0) {
+            throw new OutOfMemoryError();
+        }
+        this.keys = new ObjList<>(this.rows);
+        this.values = new ObjList<>(this.rows);
+        for (int i = 0; i < this.rows; i++) {
+            keys.add(new String[this.blocks]);
+            values.add((V[]) new Object[this.blocks]);
+        }
+        this.rowMask = this.rows - 1;
+        this.cachedGauge = cachedGauge;
+        this.hitCounter = hitCounter;
+        this.missCounter = missCounter;
+    }
+
+    @Override
+    public int capacity() {
+        return rows * blocks;
+    }
+
+    @Override
+    public void close() {
+        long freed = 0;
+        for (int i = 0; i < rows; i++) {
+            final String[] rowKeys = keys.getQuick(i);
+            final V[] rowValues = values.getQuick(i);
+            synchronized (rowKeys) {
+                for (int j = 0; j < blocks; j++) {
+                    if (rowKeys[j] != null) {
+                        rowKeys[j] = null;
+                        if (rowValues[j] != null) {
+                            rowValues[j] = Misc.freeIfCloseable(rowValues[j]);
+                            freed++;
+                        }
+                    }
+                }
+            }
+        }
+        cachedGauge.add(-freed);
+    }
+
+    @Override
+    public V poll(@NotNull CharSequence key) {
+        final int row = row(key);
+        final String[] rowKeys = keys.getQuick(row);
+        final V[] rowValues = values.getQuick(row);
+        V value = null;
+
+        synchronized (rowKeys) {
+            for (int i = 0; i < blocks; i++) {
+                if (rowKeys[i] == null) {
+                    break;
+                }
+                if (rowValues[i] != null && Chars.equals(key, rowKeys[i])) {
```

Comment:
Could you clarify the scenario? Are those queries run sequentially over a single connection or concurrently?

Current `put` scans the blocks eagerly trying to find the same key. If not found, it prefers empty blocks rather than those without a value.

Reply:
nevermind, I misread it. apologies for the confusion.

Reply:
No prob at all.